In [1]:
# Print your Emory ID here
print("Emory ID: 2550117")

Emory ID: 2550117


In [2]:
# Import the necessary libraries and load the data
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
data = pd.read_csv("/Users/jasmineliu/Downloads/QTM350/qtm350-quiz03/passwords.csv")

In [3]:
data.head()

rank  password             category  value time_unit  offline_crack_sec  \
0   1.0  password     password-related   6.91     years       2.170000e+00   
1   2.0    123456  simple-alphanumeric  18.52   minutes       1.110000e-05   
2   3.0  12345678  simple-alphanumeric   1.29      days       1.110000e-03   
3   4.0      1234  simple-alphanumeric  11.11   seconds       1.110000e-07   
4   5.0    qwerty  simple-alphanumeric   3.72      days       3.210000e-03   

   rank_alt  strength  font_size  
0       1.0       8.0       11.0  
1       2.0       4.0        8.0  
2       3.0       4.0        8.0  
3       4.0       4.0        8.0  
4       5.0       8.0       11.0

In [4]:
data["password"].dtype

dtype('O')

1. An online password attack is when someone tries to hack your account by simply trying a very large number of username/password combinations to access your account. For each `password` in our dataset, the `value` column shows the amount of time it is estimated to take an “online password attack” to hack your account. The column `time_unit` shows the units of that time value (e.g., hours, days, years, etc.). It would be much nicer if our values were of the same units so we can more easily compare the “online password guessing time” for each password. So your first task is to convert all of the values to units of hours (assume the conversion units provided below, e.g., 1 day is 24 hours, 1 week is 168 hours, etc).

In [5]:
data["time_unit"].unique()

array(['years', 'minutes', 'days', 'seconds', 'months', 'weeks', 'hours',
       nan], dtype=object)

In [6]:
# Your answer here 

units = {
    "seconds": 1 / 3600,
    "minutes": 1 / 60,
    "days": 24,
    "weeks": 168,
    "months": 720,
    "years": 8760,
}

def convert_unit_to_hours(value, time_unit):
    if time_unit in units:
        return value * units[time_unit]
    else:
        return 0
    
data["value_hours"] = data.apply(lambda index: convert_unit_to_hours(index["value"], index["time_unit"]), axis = 1)
data.head()

rank  password             category  value time_unit  offline_crack_sec  \
0   1.0  password     password-related   6.91     years       2.170000e+00   
1   2.0    123456  simple-alphanumeric  18.52   minutes       1.110000e-05   
2   3.0  12345678  simple-alphanumeric   1.29      days       1.110000e-03   
3   4.0      1234  simple-alphanumeric  11.11   seconds       1.110000e-07   
4   5.0    qwerty  simple-alphanumeric   3.72      days       3.210000e-03   

   rank_alt  strength  font_size   value_hours  
0       1.0       8.0       11.0  60531.600000  
1       2.0       4.0        8.0      0.308667  
2       3.0       4.0        8.0     30.960000  
3       4.0       4.0        8.0      0.003086  
4       5.0       8.0       11.0     89.280000

2. Are there any missing values in the dataset? If so, how many? After you have counted the, please drop them from the dataset.

In [7]:
# Your answer here
data.isna().sum()

rank                 7
password             7
category             7
value                7
time_unit            7
offline_crack_sec    7
rank_alt             7
strength             7
font_size            7
value_hours          0
dtype: int64

In [8]:
data.tail(10)

rank  password          category  value time_unit  offline_crack_sec  \
497  498.0   college         nerdy-pop   3.19    months            0.08350   
498  499.0    jester              name   3.72      days            0.00321   
499  500.0  passw0rd  password-related  92.27     years           29.02000   
500    NaN       NaN               NaN    NaN       NaN                NaN   
501    NaN       NaN               NaN    NaN       NaN                NaN   
502    NaN       NaN               NaN    NaN       NaN                NaN   
503    NaN       NaN               NaN    NaN       NaN                NaN   
504    NaN       NaN               NaN    NaN       NaN                NaN   
505    NaN       NaN               NaN    NaN       NaN                NaN   
506    NaN       NaN               NaN    NaN       NaN                NaN   

     rank_alt  strength  font_size  value_hours  
497     500.0       7.0       11.0      2296.80  
498     501.0       7.0       11.0        89.28  
499     502.0      28.0       21.0    808285.20  
500       NaN       NaN        NaN         0.00  
501       NaN       NaN        NaN         0.00  
502       NaN       NaN        NaN         0.00  
503       NaN       NaN        NaN         0.00  
504       NaN       NaN        NaN         0.00  
505       NaN       NaN        NaN         0.00  
506       NaN       NaN        NaN         0.00

In [9]:
clean_data = data.dropna()
clean_data.isna().sum()

rank                 0
password             0
category             0
value                0
time_unit            0
offline_crack_sec    0
rank_alt             0
strength             0
font_size            0
value_hours          0
dtype: int64

There are 7 missing values.

3. Which password category has the highest average strength?


In [10]:
# Your answer here
clean_data.groupby("category").agg(
    {'strength': ['mean']}
)

strength
                          mean
category                      
animal                7.551724
cool-macho            7.215190
fluffy                6.772727
food                  4.363636
name                  7.191257
nerdy-pop            14.000000
password-related      7.866667
rebellious-rude       6.727273
simple-alphanumeric   5.852459
sport                 8.000000

Nerdy-pop has the highest average strength.

4. What is the rank of the password 'qwerty'?

In [11]:
# Your answer here
index = clean_data[clean_data["password"] == 'qwerty'].index
print(index)
print(clean_data.iloc[4,:])

Index([4], dtype='int64')
rank                                 5.0
password                          qwerty
category             simple-alphanumeric
value                               3.72
time_unit                           days
offline_crack_sec                0.00321
rank_alt                             5.0
strength                             8.0
font_size                           11.0
value_hours                        89.28
Name: 4, dtype: object


The rank is 5.

5. How many passwords have more than 8 characters?

In [12]:
# Your answer here
clean_data["password"].str.len().unique()

array([8, 6, 4, 5, 7, 9])

In [13]:
(clean_data["password"].str.len() > 8).sum()

1

only one password has more than 8 characters.

6. How many passwords do not contain a number?

In [14]:
# Your answer here
def contains_numbers(password):
    if any(char.isdigit() for char in password):
        return True

clean_data["password"].apply(contains_numbers).sum()

54

In [15]:
clean_data.shape

(500, 10)

In [16]:
answer = 500-54
print(answer)

446


There are 446 passwords without a number

7. What is the average time in hours needed to crack these passwords that begin with `123`? How does this compare to the average of all passwords in the dataset?

In [17]:
# Your answer here
filter = clean_data[clean_data["password"].str.startswith('123')]
filter.head(20)
filter["value_hours"].mean()

106.95999104938272

106 hours is the average

8. What percentage of passwords are of the `simple-alphanumeric` category?

In [18]:
# Your answer here
(clean_data["category"] == 'simple-alphanumeric').sum()

61

In [19]:
answer_8 = 61/500
print(answer_8)

0.122


About 12 percent

9. How many passwords have a strength rating below 5? What is their percentage?

In [20]:
# Your answer here
(clean_data["strength"] < 5).sum()

73

In [21]:
answer_9 = 73/500
print(answer_9)

0.146


Answer: 14.6 percent

10. What is the most common password in the dataset?

In [22]:
# Your answer here
clean_data['password'].nunique()

500

Each password is unique in our dataset so there's no most common password.

11. How many unique passwords are there in each category, and what is their average length?

In [37]:
# Your answer here
grouped = clean_data.groupby('category')
unique_counts = grouped['password'].nunique()
print(unique_counts)
average_lengths = grouped['password'].apply(lambda password: password.str.len().mean())
print(average_lengths)

category
animal                  29
cool-macho              79
fluffy                  44
food                    11
name                   183
nerdy-pop               30
password-related        15
rebellious-rude         11
simple-alphanumeric     61
sport                   37
Name: password, dtype: int64
category
animal                 6.206897
cool-macho             6.253165
fluffy                 5.795455
food                   6.090909
name                   6.218579
nerdy-pop              6.633333
password-related       6.333333
rebellious-rude        6.363636
simple-alphanumeric    5.934426
sport                  6.513514
Name: password, dtype: float64


12. What is the median rank and strength for passwords of each length?

In [42]:
# Your answer here
clean_data['length'] = clean_data['password'].str.len()
clean_data.groupby('length').agg(
    {"rank" : "median",
     "strength": "median"}
)


/var/folders/sn/6ghc2n6d7qd3q_hb0b_9_cnr0000gn/T/ipykernel_20204/901567461.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  clean_data['length'] = clean_data['password'].str.len()


rank  strength
length                 
4       326.0       6.0
5       343.0       7.0
6       231.0       7.0
7       222.0       8.0
8       286.0       8.0
9        49.0       4.0

13. Add a column `strength_label` with values `weak` if strength is below 5, else `strong`. Use a lambda function.

In [44]:
# Your answer here
clean_data["strength_label"] = clean_data['strength'].apply(lambda x: 'weak' if x < 5 else 'strong')
clean_data.head()

/var/folders/sn/6ghc2n6d7qd3q_hb0b_9_cnr0000gn/T/ipykernel_20204/1601917405.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  clean_data["strength_label"] = clean_data['strength'].apply(lambda x: 'weak' if x < 5 else 'strong')


rank  password             category  value time_unit  offline_crack_sec  \
0   1.0  password     password-related   6.91     years       2.170000e+00   
1   2.0    123456  simple-alphanumeric  18.52   minutes       1.110000e-05   
2   3.0  12345678  simple-alphanumeric   1.29      days       1.110000e-03   
3   4.0      1234  simple-alphanumeric  11.11   seconds       1.110000e-07   
4   5.0    qwerty  simple-alphanumeric   3.72      days       3.210000e-03   

   rank_alt  strength  font_size   value_hours  length strength_label  
0       1.0       8.0       11.0  60531.600000       8         strong  
1       2.0       4.0        8.0      0.308667       6           weak  
2       3.0       4.0        8.0     30.960000       8           weak  
3       4.0       4.0        8.0      0.003086       4           weak  
4       5.0       8.0       11.0     89.280000       6         strong

14. Create a column `high_value` with a value of `True` if value is in the top 10% of values in the dataset.

In [45]:
# Your answer here
clean_data["high_value"] = clean_data['value'] > clean_data['value'].quantile(0.9)
clean_data.head()

/var/folders/sn/6ghc2n6d7qd3q_hb0b_9_cnr0000gn/T/ipykernel_20204/1008604279.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  clean_data["high_value"] = clean_data['value'] > clean_data['value'].quantile(0.9)


rank  password             category  value time_unit  offline_crack_sec  \
0   1.0  password     password-related   6.91     years       2.170000e+00   
1   2.0    123456  simple-alphanumeric  18.52   minutes       1.110000e-05   
2   3.0  12345678  simple-alphanumeric   1.29      days       1.110000e-03   
3   4.0      1234  simple-alphanumeric  11.11   seconds       1.110000e-07   
4   5.0    qwerty  simple-alphanumeric   3.72      days       3.210000e-03   

   rank_alt  strength  font_size   value_hours  length strength_label  \
0       1.0       8.0       11.0  60531.600000       8         strong   
1       2.0       4.0        8.0      0.308667       6           weak   
2       3.0       4.0        8.0     30.960000       8           weak   
3       4.0       4.0        8.0      0.003086       4           weak   
4       5.0       8.0       11.0     89.280000       6         strong   

   high_value  
0       False  
1        True  
2       False  
3        True  
4       False

15. Join all passwords within each `category` into a single string.

In [52]:
# Your answer here
group = clean_data.groupby('category')
joined_passwords = group['password'].agg(' '.join)
print(joined_passwords)


category
animal                 dragon monkey bigdog falcon phoenix tigers chi...
cool-macho             mustang shadow master harley hunter ranger bus...
fluffy                 love sunshine silver orange ginger summer prin...
food                   pepper cheese coffee banana butter muffin hotd...
name                   michael jennifer jordan superman thomas tigger...
nerdy-pop              starwars computer merlin matrix snoopy boomer ...
password-related       password letmein test pass access gateway plea...
rebellious-rude        biteme freedom secret whatever badboy sexsex b...
simple-alphanumeric    123456 12345678 1234 qwerty 12345 696969 abc12...
sport                  baseball football soccer hockey dallas yankees...
Name: password, dtype: object


**Bonus question 01**: What is the correlation between password length and strength? Create a scatter plot with regression line.

In [28]:
# Your answer here

**Bonus question 02**: Find all passwords with palindromes.

In [29]:
# Your answer here